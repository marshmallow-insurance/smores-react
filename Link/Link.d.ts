import { FC, MouseEventHandler, ReactNode } from 'react';
import { Icons } from 'Icon/iconsList';
type LinkTypo = 'regular' | 'small';
export type LinkProps = {
    className?: string;
    href: string;
    onClick?: MouseEventHandler;
    openInNewTab?: boolean;
    children?: ReactNode;
    download?: boolean;
    typo?: LinkTypo;
    highlight?: boolean;
    iconToRender?: Icons;
    isTrailingIcon?: boolean;
};
export declare const Link: FC<LinkProps>;
/**
 * Internal utility to override styling in other components (see Text)
 * @internal
 */
export declare const linkStyleOverride: ({ color }: {
    color: string;
}) => import("styled-components").FlattenSimpleInterpolation;
export {};
