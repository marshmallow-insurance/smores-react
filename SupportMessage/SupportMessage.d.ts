import { FC, MouseEventHandler, ReactElement } from 'react';
import { MarginProps } from '../utils/space';
type SupportMessageType = 'info' | 'fallbackStyle' | 'alert' | 'warning' | 'success';
export type SupportMessageProps = {
    className?: string;
    description: string | ReactElement;
    onClick?: MouseEventHandler;
    type: SupportMessageType;
    title?: string;
} & MarginProps;
export declare const SupportMessage: FC<SupportMessageProps>;
export {};
