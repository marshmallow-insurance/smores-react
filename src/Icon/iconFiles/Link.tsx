import React from 'react'

const Link = () => (
  <svg
    width="100%"
    height="100%"
    viewBox="0 0 24 24"
    fill="none"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M14.435 17.264l-3.142 3.142a5.45 5.45 0 01-7.699 0A5.41 5.41 0 012 16.557c0-1.454.566-2.821 1.594-3.85l3.142-3.143A.999.999 0 105.322 8.15L2.18 11.293A7.395 7.395 0 000 16.557c0 1.989.774 3.857 2.18 5.263A7.396 7.396 0 007.444 24a7.39 7.39 0 005.263-2.181l3.142-3.142a.999.999 0 10-1.414-1.413zm2.272-9.971a.999.999 0 00-1.414 0l-8 8a.999.999 0 101.414 1.414l8-8a.999.999 0 000-1.414zm5.113-5.112A7.393 7.393 0 0016.556 0a7.39 7.39 0 00-5.263 2.181L8.151 5.322a.999.999 0 101.414 1.414l3.142-3.142a5.45 5.45 0 017.699 0A5.41 5.41 0 0122 7.443a5.412 5.412 0 01-1.594 3.85l-3.142 3.143a.999.999 0 101.414 1.414l3.142-3.143A7.395 7.395 0 0024 7.443a7.388 7.388 0 00-2.18-5.262z"
      fill="currentColor"
    />
  </svg>
)

export default Link
