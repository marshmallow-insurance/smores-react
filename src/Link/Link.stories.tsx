import React from 'react'
import { Text } from '../Text'

import { Link, LinkProps } from './Link'

export default {
  title: 'Link',
  component: Link,
  argTypes: { onClick: { action: 'clicked' } },
}

const linkArgs = {
  href: 'https://www.google.com',
  openInNewTab: true,
  children: 'Google link',
  typo: 'regular',
}

export const Playground = Link.bind({})
Playground.args = linkArgs

export const Highlighted = Link.bind({})
Highlighted.args = {
  ...linkArgs,
  highlight: true,
}

export const LeadingIcon = Link.bind({})
LeadingIcon.args = {
  ...linkArgs,
  iconToRender: 'new-window',
  isTrailingIcon: false,
}

const ParagraphDemo = (props: LinkProps) => (
  <Text tag="p" color="apple">
    Lorem Ipsum is simply dummy text of the printing and typesetting industry,
    more details <Link {...props} />. Lorem Ipsum has been the industrys
    standard dummy text ever since the 1500s.
  </Text>
)

export const InParagraph = ParagraphDemo.bind({})
InParagraph.args = {
  ...linkArgs,
  href: 'https://en.wikipedia.org/wiki/Lorem_ipsum',
  children: 'here',
}
