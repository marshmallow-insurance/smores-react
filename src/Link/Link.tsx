import { darken } from 'polished'
import React, { FC, MouseEventHandler, ReactNode } from 'react'
import styled, { css } from 'styled-components'

import { Icon } from '../Icon'
import { Icons } from '../Icon/iconsList'
import { theme } from '../theme'
import { focusOutline } from '../utils/focusOutline'

type LinkTypo = 'regular' | 'small'

export type LinkProps = {
  className?: string
  href: string
  onClick?: MouseEventHandler
  openInNewTab?: boolean
  children?: ReactNode
  download?: boolean
  typo?: LinkTypo
  highlight?: boolean
  iconToRender?: Icons
  isTrailingIcon?: boolean
}

export const Link: FC<LinkProps> = ({
  className,
  href,
  onClick,
  openInNewTab,
  download,
  children,
  typo = 'regular',
  highlight = false,
  iconToRender = openInNewTab ? 'new-window' : undefined,
  isTrailingIcon = true,
}) => {
  return (
    <LinkWrapper
      href={href}
      className={className}
      onClick={onClick}
      download={download}
      $typo={typo}
      $highlight={highlight}
      {...(openInNewTab && {
        rel: 'noopener noreferrer',
        target: '_blank',
      })}
    >
      {iconToRender && !isTrailingIcon && (
        <Icon
          mt={{ custom: '3px' }}
          mr={{ custom: '4px' }}
          render={iconToRender}
          size={typo === 'regular' ? 14 : 12}
          color={highlight ? 'lollipop' : 'liquorice'}
        />
      )}
      {children}
      {iconToRender && isTrailingIcon && (
        <Icon
          mt={{ custom: '3px' }}
          ml="8px"
          render={iconToRender}
          size={typo === 'regular' ? 14 : 12}
          color={highlight ? 'lollipop' : 'liquorice'}
        />
      )}
    </LinkWrapper>
  )
}

const LinkWrapper = styled.a<{ $typo: LinkTypo; $highlight: boolean }>(
  ({ $typo, $highlight }) => css`
    ${focusOutline()}
    display: inline-flex;
    flex-direction: row;

    ${$typo === 'regular' &&
    css`
      font-size: 16px;
      line-height: 20px;
    `}

    ${$typo === 'small' &&
    css`
      font-size: 14px;
      line-height: 20px;
    `}

      font-weight: ${theme.font.weight.medium};
    text-decoration: underline;
    color: ${$highlight ? theme.colors.lollipop : theme.colors.liquorice};

    background: none;
    cursor: pointer;

    &:hover {
      color: ${theme.colors.sesame};

      path {
        fill: ${theme.colors.sesame};
      }
    }

    &:active {
      color: ${theme.colors.liquorice};

      path {
        fill: ${theme.colors.liquorice};
      }
    }
  `,
)

/**
 * Internal utility to override styling in other components (see Text)
 * @internal
 */
export const linkStyleOverride = ({ color }: { color: string }) => css`
  & ${LinkWrapper} {
    color: ${color};

    path {
      fill: ${color};
    }

    &:hover {
      color: ${darken(0.1, color)};

      path {
        fill: ${darken(0.1, color)};
      }
    }
  }
`
