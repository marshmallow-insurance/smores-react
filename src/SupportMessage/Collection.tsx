import React, { FC } from 'react'
import styled from 'styled-components'

import { Text } from '../Text'
import { Box } from '../Box'
import { SupportMessage, SupportMessageProps } from './SupportMessage'

const variationList: SupportMessageProps[] = [
  {
    type: 'info',
    title: 'Additional information',
    description: 'Information needs to be communicated to the user.',
  },
  {
    type: 'fallback',
    title: 'Additional information',
    description: 'Information needs to be communicated to the user.',
  },
  {
    type: 'alert',
    title: 'Potential issue',
    description: 'The user should be aware and possibly exercise caution.',
  },
  {
    type: 'warning',
    title: 'Error encountered',
    description: "An error has occurred, here's feedback on how to proceed.",
  },
  {
    type: 'success',
    title: 'Error encountered',
    description: "An error has occurred, here's feedback on how to proceed.",
  },
]

export const CollectionPage: FC = () => {
  return (
    <Box flex direction="column">
      {variationList.map(({ type, title, description }) => {
        return (
          <Box key={type} flex direction="column" mb="32px">
            <Title tag="span" typo="header-medium">
              Type: {type}
            </Title>
            <Box flex>
              <Label tag="span" typo="header-small">
                Without title
              </Label>
              <SupportMessage type={type} description={description} />
            </Box>
            <Box flex mt="8px">
              <Label tag="span" typo="header-small">
                With title
              </Label>
              <SupportMessage
                type={type}
                title={title}
                description={description}
              />
            </Box>
            <Box flex mt="8px">
              <Label tag="span" typo="header-small">
                Clickable
              </Label>
              <SupportMessage
                type={type}
                title={title}
                description={description}
                onClick={() => alert('Clicked!')}
              />
            </Box>
          </Box>
        )
      })}
    </Box>
  )
}

const LABEL_WIDTH = '140px'

const Title = styled(Text)`
  margin-bottom: 12px;
  font-size: 18px;
  line-height: 20px;
  margin-left: ${LABEL_WIDTH};
`

const Label = styled(Text)`
  margin-bottom: 4px;
  font-size: 14px;
  width: ${LABEL_WIDTH};
  flex-shrink: 0;
`
