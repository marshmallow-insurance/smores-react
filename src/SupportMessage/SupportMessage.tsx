import React, { FC, ReactElement } from 'react'
import styled, { css } from 'styled-components'
import { lighten } from 'polished'

import { Box } from '../Box'
import { Icon } from '../Icon'
import { theme, Color } from '../theme'
import { Text } from '../Text'

type StylesItem = {
  iconColor: Color
  backgroundColor: string
  icon: string
}

const styles: Record<SupportMessageType, StylesItem> = {
  info: {
    iconColor: 'secondary',
    backgroundColor: theme.colors.background,
    icon: 'info',
  },
  'info-outline': {
    iconColor: 'secondary',
    backgroundColor: theme.colors.white,
    icon: 'info',
  },
  alert: {
    iconColor: 'agentWarning',
    backgroundColor: theme.colors.bgSecondary,
    icon: 'alert',
  },
  warning: {
    iconColor: 'error',
    backgroundColor: lighten(0.45, theme.colors.error),
    icon: 'warning',
  },
}

type SupportMessageType = 'info' | 'info-outline' | 'alert' | 'warning'

export type SupportMessageProps = {
  className?: string
  description: string | ReactElement
  type: SupportMessageType
  title?: string
}

export const SupportMessage: FC<SupportMessageProps> = ({
  className,
  description,
  type = 'info',
  title,
}) => (
  <Wrapper className={className} type={type}>
    <IconWrapper>
      <Icon
        size={24}
        render={styles[type].icon}
        color={styles[type].iconColor}
      />
    </IconWrapper>
    <Box flex direction="column" ml="16px">
      {title && <Title>{title}</Title>}
      <Description tag="p" typo="base">
        {description}
      </Description>
    </Box>
  </Wrapper>
)

interface IWrapper {
  type: SupportMessageType
}

const IconWrapper = styled(Box)`
  align-self: flex-start;
`

const Wrapper = styled.div<IWrapper>(
  ({ type }) => css`
    align-items: center;
    background-color: ${styles[type].backgroundColor};
    ${type === 'info-outline' && `border: 1px solid ${theme.colors.secondary}`};
    border-radius: 8px;
    padding: 16px;
    display: flex;
    width: 100%;
  `,
)

const Title = styled.p`
  font-size: 16px;
  font-weight: ${theme.font.weight.medium};
  color: ${theme.colors.secondary};
  line-height: 20.8px;
  margin-bottom: 4px;
`

const Description = styled(Text)`
  color: ${theme.colors.secondary};
  font-size: 14px;
  line-height: 20px;
`
