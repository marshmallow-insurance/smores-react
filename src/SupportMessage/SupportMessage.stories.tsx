import React from 'react'
import { SupportMessage, SupportMessageProps } from './SupportMessage'
import { Link } from '../Link'
import { CollectionPage } from './Collection'

export default {
  title: 'SupportMessage',
  component: SupportMessage,
}

const Template = (props: SupportMessageProps) => (
  <SupportMessage {...props}>Support message for customer</SupportMessage>
)

export const Default = Template.bind({})

const supportMessageArgs: SupportMessageProps = {
  type: 'info',
  title: undefined,
  description: 'Some description text',
}

Default.args = supportMessageArgs

export const WithCustomDescription = Template.bind({})

WithCustomDescription.args = {
  type: 'info',
  title: 'A SupportMessage using the Link component',
  description: (
    <>
      Some text rendered using a <Link href={''}>Link</Link>
    </>
  ),
}

export const Clickable = Template.bind({})

Clickable.args = {
  type: 'info',
  title: 'An interactive SupportMessage',
  description: 'Click me!',
  onClick: () => alert('Clicked!'),
}

export const Collection = CollectionPage.bind({})
