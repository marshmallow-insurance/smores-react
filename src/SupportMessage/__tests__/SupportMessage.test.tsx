import { render, screen } from '@testing-library/react'

import React from 'react'
import { Link } from '../../Link'
import { SupportMessage, SupportMessageProps } from '../SupportMessage'

const supportMessageProps: SupportMessageProps = {
  description: 'Type info support message',
  type: 'info',
}

describe('SupportMessage component', () => {
  const { findByText } = screen
  it('renders the correct content for type info', async () => {
    render(<SupportMessage {...supportMessageProps} />)

    expect(await findByText('Type info support message')).toBeTruthy()
  })

  it('renders the correct content for type info with title', async () => {
    render(<SupportMessage {...supportMessageProps} title="Info title" />)

    expect(await findByText('Type info support message')).toBeTruthy()
    expect(await findByText('Info title')).toBeTruthy()
  })

  it('Renders a Link within the description', () => {
    const { container } = render(
      <>
        Some text rendered with a <Link href={''}>Link</Link>{' '}
      </>,
    )
    expect(container.firstChild).toMatchSnapshot()
  })
})
