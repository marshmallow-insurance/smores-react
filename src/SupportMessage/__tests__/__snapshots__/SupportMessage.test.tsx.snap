// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SupportMessage component > Renders a Link within the description 1`] = `Some text rendered with a `;
